// graphql-let/dist/lib/create-codegen-opts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const graphql_1 = require("graphql");
const cli_1 = require("@graphql-codegen/cli");
function isURL(path) {
    try {
        new URL(path);
        return true;
    }
    catch (e) {
        return false;
    }
}
function loadSchema(pointer) {
    return cli_1.CodegenContext.prototype.loadSchema.call({ getConfig: () => ({}) }, pointer);
}
async function generateSchema(path, cwd) {
    const schemaPointer = isURL(path) || path_1.isAbsolute(path) ? path : path_1.join(cwd, path);
    const loadedSchema = await loadSchema(schemaPointer);
    return graphql_1.parse(graphql_1.printSchema(loadedSchema));
}
async function createCodegenOpts(config, cwd) {
    return {
        config: {
            withHOC: false,
            withHooks: true,
        },
        schema: await generateSchema(config.schema, cwd),
        plugins: config.plugins.map(name => ({ [name]: {} })),
        pluginMap: config.plugins.reduce((acc, name) => {
            return { ...acc, [name]: require(`@graphql-codegen/${name}`) };
        }, {}),
    };
}
exports.default = createCodegenOpts;
